### PR TITLE
[AutoDiff] Fix falsely triggered precondition in Array.DifferentiableView.+ and .move

### PR DESCRIPTION
**Pull request by vojtamolda**
Dear everyone,

This PR adds a fix and tests that exercise falsely triggered preconditions in the `Array.DifferentiableView.+` and `.move(by:)` methods in the standard library.

The fix implements a correct handling of `.zero` tangent vector, i.e. an empty `[]` array, in the back-propagation pass of the aforementioned methods. It avoids count mis-match when the de-concatenated gradient passed to `+(...)` pullback happens to be zero. I've also bundled-in a small rewording of the precondition error messages that now includes the invalid count besides the expected one.

I didn't bother fixing nor testing the forward-propagation pass `Array.DifferentiableView._jvpContatenate` since the plan is to remove these from the repository anyway. Is this assumption correct?

Resolves [SR-14297](https://bugs.swift.org/browse/SR-14297).

Cheers,

  Vojta

@rxwei or whoever happens to be the reviewer. This is my first PR, so I apologize for violating any standards upfront. I did my best to follow the [Getting started](https://github.com/apple/swift/blob/main/docs/HowToGuides/GettingStarted.md#running-tests) and [First PR](https://github.com/apple/swift/blob/main/docs/HowToGuides/FirstPullRequest.md) guides but some things may have slipped through.

@vguerra I hope you like the improved precondition messages that now contain the invalid lengths! Thanks for the suggestion ;)

### CODE REVIEW

File: stdlib/public/Differentiation/ArrayDifferentiation.swift [hunk 1]
```diff
@@ -218,10 +219,13 @@ extension Array where Element: Differentiable {
...
     func pullback(_ v: TangentVector) -> (TangentVector, TangentVector) {
       precondition(
-        v.base.count == lhs.count + rhs.count, """
-          Tangent vector with invalid count; expected to equal the sum of \
-          operand counts \(lhs.count) and \(rhs.count)
+        v.base.isEmpty || v.base.count == lhs.count + rhs.count, """
+          Tangent vector with invalid count \(v.base.count); expected to \
+          equal the sum of operand counts \(lhs.count) and \(rhs.count)
           """)
+      if v.base.isEmpty {
```
Comment (rxwei) [inline stdlib/public/Differentiation/ArrayDifferentiation.swift:226]: If you move the `if` block before the `precondition`, the precondition won't need to be changed.

File: stdlib/public/Differentiation/ArrayDifferentiation.swift [hunk 2]
```diff
@@ -74,11 +74,12 @@ where Element: Differentiable {
...
-      base.count == offset.base.count, """
+      offset.base.isEmpty || base.count == offset.base.count,
+        """
         Count mismatch: \(base.count) ('self') and \(offset.base.count) \
         ('direction')
         """)
-    for i in base.indices {
+    for i in offset.base.indices {
       base[i].move(by: offset.base[i])
     }
```
Comment (rxwei) [inline stdlib/public/Differentiation/ArrayDifferentiation.swift:84]: I'd suggest adding an early return before `precondition` and reverting the changes to the precondition.
```swift
if offset.base.isEmpty {
  return
}
```

**Comment by rxwei** [comment]
> I didn't bother fixing nor testing the forward-propagation pass `Array.DifferentiableView._jvpContatenate` since the plan is to remove these from the repository anyway. Is this assumption correct?

Yep, this is correct.

> @rxwei or whoever happens to be the reviewer. This is my first PR, so I apologize for violating any standards upfront. I did my best to follow the Getting started and First PR guides but some things may have slipped through.

Thanks for the contribution. It already looks great!

File: stdlib/public/Differentiation/ArrayDifferentiation.swift (hunk 1 above)
Comment (vojtamolda) [reply stdlib/public/Differentiation/ArrayDifferentiation.swift]: Ok.

File: stdlib/public/Differentiation/ArrayDifferentiation.swift (hunk 2 above)
Comment (vojtamolda) [reply stdlib/public/Differentiation/ArrayDifferentiation.swift]: Sure thing.

**Comment by vojtamolda** [comment]
Thanks for the comments and the review, @rxwei.  I don't have a problem with any of them and the PR is already updated.

Now, a noob question. Can I trigger CI or does it run automatically?

**Comment by rxwei** [comment]
Thanks! CI can only be triggered by contributors with commit access. I'll trigger it for you.

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test and merge

**Pull request merged**